src/graph/graphcontext.c

```diff
@@ -45,40 +37,125 @@ GraphContext *GraphContext_New(RedisModuleCtx *ctx, const char *graphname,
 
 	gc->string_mapping = array_new(char *, 64);
 	gc->attributes = raxNew();
-
+	Graph_SetMatrixPolicy(gc->g, SYNC_AND_MINIMIZE_SPACE);
 	QueryCtx_SetGraphCtx(gc);
 
-	// Set and close GraphContext key in Redis keyspace
-	RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+	return gc;
+}
 
-	GraphContext_RegisterWithModule(gc);
+/* This method tries to get a graph context, and if it does not exists, create a new one.
+ * The try-get-create flow is done when module global lock is acquired, to enforce consistency
+ * while BGSave is called. */
+static GraphContext *_GraphContext_Create(RedisModuleCtx *ctx, const char *graphname,
+										  int readWriteFlag, size_t node_cap, size_t edge_cap) {
+	// Create and initialize a graph context.
+	GraphContext *gc = _GraphContext_New(graphname, node_cap, edge_cap);
+	RedisModuleString *rs_name = RedisModule_CreateString(ctx, graphname, strlen(graphname));
+
+	// Acquire lock.
+	assert(pthread_mutex_lock(&_module_mutex) == 0);
+	// Open key with requested permission.
+	RedisModuleKey *key = RedisModule_OpenKey(ctx, rs_name, readWriteFlag);
+	if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+		// In case of non existing value, open key for write mode.
+		key = RedisModule_OpenKey(ctx, rs_name, REDISMODULE_WRITE);
```

Comment:
In the worst case of a read query against a non-existent key, we'll call `RedisModule_OpenKey` 3 times without corresponding closes before actually creating the graph, which feels a bit wonky?